### PR TITLE
refactor(core): refactor API of Tokens

### LINKED ISSUE #1933: Revisit tokens APIs
Tokens are a queer beast. We should look into their APIs and make sure they are not easy to abuse.

Thoughts:

1. `Token.unresolved` instead of `cdk.unresolve`
2. Hide `resolve` so that it doesn't appear in public API
3. Factory methods for specific use cases instead of `new Token`. For example `Token.lazy(() => )`
4. Tokens are conceptually coupled to synthesis and now also must be anchored by a construct scope ("lazy" means - resolve during synthesis), maybe we can simply move all token facilities to `Construct.node`

### PR DESCRIPTION
**Pull request by rix0rrr**
Tokens are now represented by the interface `IResolvable`. The class `Token`
still exists, but it is used to hold static routines for token encoding
(`Token.asString()`).

Actual token use is split into 2 categories:

- *Intrinsics*, represented by static methods on `Token`: these are to represent values
  that will be understood by the deployment language formalism (e.g.
  CloudFormation), and can be used to escape language type checking.
- *Lazy values*, represented by a `Lazy` class with static factories:
  these will be used to represent type-checked but lazily-produced
  values (evaluated at synthesis time).

In order to be JSII-compatible (which does not currently support
lambdas), `Lazy.stringValue()` et. al. take an interface with a single
method instead of a function.

Also changed in this commit: shoring up property names for encoded
tokens in classes like `CfnParameter` and `CfnElement`.

* `.stringValue` => `.valueAsString`, etc so `.value`, `.valueAsString`,
  `.valueAsList` etc. group together in autocomplete.
* `.ref` now returns a Token, `.refAsString` returns the stringified
  version that token.

To match the other standard the latter should actually be `.refAsString`
but `.ref` is used in so many places that this might be uncomfortable?

Fixes #1933.

BREAKING CHANGES:

* `Token` can no longer be instantiated. You probably want to use 
  `Lazy.stringValue` and others, or `Token.asString()` and others.
* `Token.isToken()`/`Token.unresolved()` => `Token.isUnresolved()`.
* `CfnOutput`: remove `.ref`, since they can't be referenced anyway.
* `CfnParameter`: rename `.stringValue` => `.valueAsString` (and
  similar for lists and numbers).
* `.ref` now returns an `IToken`, use `.refAsString` to get a
  string-encoded Token.
* `TokenMap` is now no longer exported, some parts of its functionality
  are available through the static `Tokenization` class.
* `IResolvedValuePostProcessor` => `IResolvableWithPostProcess`.

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [ ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/cdk/lib/lazy.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  /**
+   * If the produced value is an array and it is empty, return 'undefined' instead
+   *
+   * @default false
+   */
+  readonly omitEmptyArray?: boolean;
+}
+
+/**
+ * Lazily produce a value
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/lazy.ts]: To me, `Lazy` means that we'll spare a bunch of expensive compute if some value is never used. That's not what we do here, more like deferring the compute to a later stage (the `prepare` or `synth` phase, I guess?).

I don't think `Lazy` is so fundamentally wrong that this is a deal breaker to me, however I kinda feel like something like `Deferred` or `SynthesisTime` would be a more accurate description of what happens.

File: packages/@aws-cdk/cdk/lib/lazy.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
+import { IResolveContext, Token } from "./token";
+
+/**
+ * Interface for lazy string producers
+ */
+export interface IStringValue {
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/lazy.ts]: Is this not an `IStringValueProducer`? The same applies too... the other `I*Value` interfaces here.

File: packages/@aws-cdk/cdk/lib/lazy.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/**
+ * Interface for lazy list producers
+ */
+export interface IListValue {
+  /**
+   * Produce the list value
+   */
+  produce(context: IResolveContext): string[] | undefined;
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/lazy.ts]: That makes it an `IStringListValue[Producer]`...

File: packages/@aws-cdk/cdk/lib/lazy.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/**
+ * Lazily produce a value
+ *
+ * Can be used to return a string, list or numeric value whose actual value
+ * will only be calculated later, during synthesis.
+ */
+export class Lazy {
+  public static stringValue(producer: IStringValue, options: LazyStringValueOptions = {}) {
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/lazy.ts]: I kinda would like tho have the return types of those be spelled out on the signature. But I'm cool if you choose not to.

File: packages/@aws-cdk/aws-apigateway/lib/deployment.ts [hunk 5]
```diff
@@ -76,7 +76,7 @@ export class Deployment extends Resource {
     }
 
     this.api = props.api;
-    this.deploymentId = new Token(() => this.resource.deploymentId).toString();
+    this.deploymentId = Lazy.stringValue({ produce: () => this.resource.deploymentId });
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/lib/deployment.ts]: Can we shorten? `Lazy.string` conveys the exact same meaning with less words...

File: tools/cfn2ts/lib/genspec.ts [hunk 6]
```diff
@@ -174,6 +174,8 @@ export function attributeDefinition(resourceName: CodeName, attributeName: strin
   let attrType: string;
   if ('PrimitiveType' in spec && spec.PrimitiveType === 'String') {
     attrType = 'string';
+  } else if ('PrimitiveType' in spec && spec.PrimitiveType === 'Integer') {
```
Comment (eladb) [inline tools/cfn2ts/lib/genspec.ts]: Thank you!

File: packages/@aws-cdk/cdk/lib/token.ts [hunk 7]
```diff
@@ -19,39 +46,58 @@ export const RESOLVE_METHOD = 'resolve';
...
   /**
-   * Returns true if obj is a token (i.e. has the resolve() method or is a
-   * string or array which includes token markers).
+   * Returns true if obj is an IToken object.
    *
    * @param obj The object to test.
    */
-  public static isToken(obj: any): boolean {
-    return unresolved(obj);
+  public static isToken(obj: any): obj is IToken {
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/token.ts]: I feel `Token.isToken(x)` should return true if `x` is a token or a token encoded as a primitive (list/string/number). 
The test of whether something implements `IToken` should probably not even be a public API.

File: packages/@aws-cdk/cdk/lib/fn.ts [hunk 8]
```diff
@@ -292,14 +293,14 @@ export class Fn {
    * @returns a token represented as a string array
    */
   public valueOfAll(parameterType: string, attribute: string): string[] {
-    return new FnValueOfAll(parameterType, attribute).toList();
+    return Token.encodeAsList(new FnValueOfAll(parameterType, attribute));
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/fn.ts]: At this point I don't think we need all the FnXxx private classes any more. Let's just embed the intrinsic function directly.

File: packages/@aws-cdk/cdk/lib/intrinsic.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  /**
+   * A hint for the Token's purpose when stringifying it
+   */
+  readonly displayHint?: string;
+}
+
+/**
+ * Token subclass that represents values intrinsic to the target document language
+ */
+export class Intrinsic extends Token {
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/intrinsic.ts]: TBH I am not sure why this is needed? Kind of feels like `encodeAsXxx` should accept `any` which can be either an `IToken` or some primitive values that will be passed through as a token.

I also don't like the CFN semantics that "seeped through" to this layer. If you insist on keeping this class (again, I don't think it's really needed), then I would call it `AnyValue` or `AnyToken` or something like that.

File: packages/@aws-cdk/aws-apigateway/lib/deployment.ts (hunk 5 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-apigateway/lib/deployment.ts]: `string` will be a keyword  in nearly all languages.

File: packages/@aws-cdk/aws-ec2/lib/security-group-rule.ts [hunk 10]
```diff
@@ -239,7 +239,7 @@ export class UdpPort implements IPortRange {
  * A UDP port range
  */
 export class UdpPortRange implements IPortRange {
-  public readonly canInlineRule = !Token.isToken(this.startPort) && !Token.isToken(this.endPort);
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-ec2/lib/security-group-rule.ts]: But it does  something completely different than `Array.isArray`, `Construct.isConstruct`, and every other `Class.isClass` method we've added anywhere else.

File: packages/@aws-cdk/cdk/lib/fn.ts (hunk 8 above)
Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/fn.ts]: Maybe, but they're all private so not really hurting.

File: packages/@aws-cdk/cdk/lib/intrinsic.ts (hunk 9 above)
Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/intrinsic.ts]: The word  "intrinsic" happens to exist in CloudFormation already, but I actually think it makes sense to have this as a concept in CDK: a value that makes sense in the target  document language, which might not strictly adhere to the typing of the host language. The  document language might be CFN, might be Google Cloud Manager,  might be Kubernetes YAML, might be anything. There will still be value in the "intrinsic" concept, and nobody is forcing us to call it something else than it is called  in CloudFormation. 

As a concept name, I like "intrinsic" better than "AnyValue". 

I also don't see why we need to bind the representation of the "AnyValue" and the encoding together. To me, it feels better to make them explicitly two operations: first represent a document language-level value, and *maybe* additionally encode it.

Users won't need to see any of this anyway, they will have  the necessary operations pre-wrapped, and only use the `Lazy` constructors.

**Comment by rix0rrr** [comment]
TODO (just realized):

- `Lazy.anyValue()` needs to return an  `IToken`, otherwise jsii  languages won't be able to pass the result in anywhere (`any` escapes typesystem is only valid in TypeScript).
- `cfn2ts` and all other places that say they accept a `Token` need to accept an `IToken`.

**Pull request merged**